### DISCUSSION TITLE
[Feature] Auto Image Rotation Detection

### DISCUSSION THREAD

**Discussion by Tyree** [immich-app/immich#10136] [Feature Request]
### I have searched the existing feature requests to make sure this is not a duplicate request.

- [X] Yes

### The feature

Certain apps* have a feature that detects rotated images and curates them for the user to choose to rotate them as suggested or not. Similar to how Immich now detects duplicates, I'd love to see it be able to automagically find these rotated images!
Thanks!

#1695 is similar, but it's for a manual image rotation tool and not for auto detection. Perhaps it could just be extended to include an auto-detect functionality.

### Platform

- [X] Server
- [X] Web
- [X] Mobile

**Comment by NicholasFlamy** [comment]
A message with a suggestion for this, including a link to an existing AI model for this:
https://discord.com/channels/979116623879368755/1129456657995403294/1246841616191324160

Link to existing AI model:
https://github.com/pidahbus/deep-image-orientation-angle-detection

**Reply by hexxone** [reply]
+1 for the idea
Thanks for reposting this. Would be very cool to see!

**Comment by Chuckame** [comment]
If we don't want to touch the original photo, we could generate related xmp sidecar using the `tiff:Orientation` property coming from [Adobe xmp specs](https://developer.adobe.com/xmp/docs/XMPNamespaces/tiff/)

**Reply by NicholasFlamy** [reply]
Pretty sure external libraries can already write sidecar XMP files with image editing and rotating if you have read-only disabled.

**Reply by dionorgua** [reply]
It looks like Immich doesn't support `tiff:Orientation`. 
It doesn't work for me: 
https://github.com/immich-app/immich/issues/15549
https://github.com/immich-app/immich/discussions/15550

**Comment by parsapoorsh** [comment]
i had this problem, but i couldn't find a lightweight model to do rotation detection, so i **glued** one together:
https://github.com/parsapoorsh/resnet-ixion

it's size is about ~220 MB, and it has a great performance for images that have an object or face in them (like the images in [demo instance](https://demo.immich.app/photos))

but it can't detect rotation of scanned papers or an empty wall and stuff like that

**Reply by NicholasFlamy** [reply]
Can you export it to ONNX? I see that PyTorch has docs on it: https://pytorch.org/tutorials//beginner/onnx/export_simple_model_to_onnx_tutorial.html

**Reply by parsapoorsh** [reply]
yes, https://github.com/parsapoorsh/resnet-ixion/releases/download/0.0.0/resnet152_ixion_e1-7c6a5234.onnx

and the script for testing it:
https://github.com/parsapoorsh/resnet-ixion/blob/master/run_onnx.py

`$ python run_onnx.py resnet152_ixion_e1-7c6a5234.onnx images/image_2.jpg`

**Reply by parsapoorsh** [reply]
here is an updated version:
https://github.com/parsapoorsh/resnet-ixion/releases/tag/1.0.0
it's good enough for me

**Reply by NicholasFlamy** [reply]
> but it can't detect rotation of scanned papers or an empty wall and stuff like that

Maybe I'll try retraining it if I can get ROCm working with PyTorch.

**Reply by NicholasFlamy** [reply]
> yes, https://github.com/parsapoorsh/resnet-ixion/releases/download/0.0.0/resnet152_ixion_e1-7c6a5234.onnx
> 
> and the script for testing it: https://github.com/parsapoorsh/resnet-ixion/blob/master/run_onnx.py
> 
> `$ python run_onnx.py resnet152_ixion_e1-7c6a5234.onnx images/image_2.jpg`

This works.

Because I'm not experienced enough I was running into issues running ROCm (installing onnxruntime with ROCm is a bit harder, unrelated to your work) but it works great on CPU.

**Reply by NicholasFlamy** [reply]
We'd need a check to handle cases like this where it's a logo that seems fine in any orientation. Something like if all angles are within 15% then do nothing.
The first image is the original logo and the second was scaled to 200 px.
<img width=40% alt="advanced-reboot-logo-high-res-rotation-test" src="https://github.com/user-attachments/assets/1c5d5863-3caa-44a2-af4c-7fb76158199d">
<img width=40% alt="advanced-reboot-logo-200px-rotation-test" src="https://github.com/user-attachments/assets/5deca40c-c82e-4d24-88c7-b27d78c30fbd">

**Reply by NicholasFlamy** [reply]
Holy sh*t I got ROCm working.

**Reply by parsapoorsh** [reply]
> We'd need a check to handle cases like this where it's a logo that seems fine in any orientation. Something like if all angles are within 15% then do nothing.

i agree, i stumbled upon the following images, and having a diff check of something like 15% or averaging all of scores fixes this issue:

<img width=60% alt="Untitled-resized" src="https://github.com/user-attachments/assets/62df6546-3002-4b5e-b142-3ce63a44e8c1">
<img width=60% alt="Untitled1-resized" src="https://github.com/user-attachments/assets/256ede8e-c3fd-41fb-8ae5-439e53baba14">


that said, it's not supposed to check for rotation of a **LOGO**, it's trained to find rotation of normal (not spam) images taken from a camera, which would have recognizable objects in them

<img width=60% alt="aaaa-resized" src="https://github.com/user-attachments/assets/78f8a439-00b2-4414-adfc-6a6b77ac2282">

**Reply by NicholasFlamy** [reply]
> that said, it's not supposed to check for rotation of a **LOGO**, it's trained to find rotation of normal (not spam) images taken from a camera, which would have recognizable objects in them

People using immich will likely have other pictures in their libraries, so that's gonna come up. That's why I'm thinking of retraining your trained model with a bunch of documents and stuff.

**Reply by parsapoorsh** [reply]
> retraining your trained model with a bunch of documents and stuff.

checkout these datasets:
https://universe.roboflow.com/browse/documents
https://sg-vilab.github.io/event/warpdoc/

also it's possible to synthetically generate data from these. for example putting them in a "screenshot with a camera" format, and other stuff.

but im also thinking about having multiple orientation detection models, fine-tuned for specific type of images.
for example using an OCR model to detect rotation if the main one fails and gets low score.
this doesn't seem to be computationally expensive, because it only calculates when the main model fails.

<img width=60% alt="dddddddd-resized" src="https://github.com/user-attachments/assets/76a76451-b5b6-43a0-8a1b-cb88c5768bb7">

**Reply by NicholasFlamy** [reply]
> but im also thinking about having multiple orientation detection models, fine-tuned for specific type of images.

If you overcomplicate it, then it'll be harder to maintain in immich.

> for example using an OCR model to detect rotation if the main one fails and gets low score.

OCR is something that seems like a lot of people want in immich so maybe that can somehow be integrated if it's added.

**Reply by parsapoorsh** [reply]
i'm downloading `ImageNet 2017` to train a better version. the current version generally has good accuracy, but it struggles with edge cases.

**Reply by hexxone** [reply]
fyi, I couldnt wait and wanted to try your model as well, so I forked your repo and added a CLI tool for batch processing:
<https://github.com/hexxone/resnet-ixion>

Disclaimer: I'm not really confident in my python and AI knowledge, but I ran this tool on my external library of photos I was planning to import into immich and it seems to yield good results:
```
Completed: 100% (10681/10681)
Total time: 7m 25s
==== batch processing complete ====
Processed: 10681 images
Had Errors: 24 images
Inconsistent: 644 images
Corrected 515 images
```
the errors are mainly caused by video files, because I didnt bother to exclude them yet.
the inconsistent amount could probably be brought down further as well, I just didnt know the best way/math for doing so..
Maybe this helps in some way :) Cheers

**Reply by parsapoorsh** [reply]
hi, sorry for the delay and not being around, im currently not in a location with access to my PC and my Laptop isn't a gaming one, so i can't continue doing the AI stuff. i will resume soon.

> Disclaimer: I'm not really confident in my python and AI knowledge

im not an AI expert ether. im new to this, but doing this stuff has become easier then ever and there are great resources around to learn from.

> added a CLI tool for batch processing

i made a GUI for batch processing ([gui.py](https://github.com/parsapoorsh/resnet-ixion/blob/master/gui.py)), which shows the images that it's not confident enough (score lower then 98%), to be manually decided.

![skate](https://github.com/user-attachments/assets/11495fe6-fbdf-4d62-8d97-07e9cb9b22f6)

through this, i found out the model isn't reliable enough to be a part of immich yet, most of the time it performs really good, but for some images it flops badly.

edge cases i found where model is not reliable:
1. if there aren't enough obvious objects
2. low brightness
3. the image is near a 45 deg angle
4. text documents

possible solutions:
1) im thinking to solve the "not enough obvious objects" problem, instead of **fine-tunning** `resnet152`, **train** `efficientnet b3` on `ImageNet 2017` + `MS COCO`, but it's gonna take like a week per epoch on my GTX 1660 super

2.3. for the low brightness and bad angles, i can randomly rotate the images 11.25 deg, make them dark and stuff like that (data augmentation), which needs more then 1 epoch for the training to have the optimal results (time-consuming)

4. text documents are a bit tricky, we need the model to detect a text's angle, (not OCR, just angle detection).
but i have no idea how make it to do so.
maybe it's possible by overlaying some random text on training images, but i think that going to harm the main normal image rotation detection.
also the text needs to be in multiple visually different writing-languages, like english, russian, arabic, hindi, etc, and in different fonts, which just complicate things for document angle detection.
because of all of these complications, i believe documents rotation detection and normal image detection should have different models.

video rotation detection:
> the errors are mainly caused by video files

doing this for videos is easy, for speed reasons just extract the **keyframes** (related to video compression), exclude the visually similar ones, and run the model on reaming ones, average the scores.

basically this ffmpeg command:

```bash
ffmpeg -skip_frame nokey -i input.mp4 -vf "mpdecimate,select='eq(pict_type,I)*gt(scene,0.35)'" -vsync vfr output_%04d.jpg
```
here is the result on Bad Apple (3m40s video, 78 pictures output):

![ccccccc-resize](https://github.com/user-attachments/assets/81565b56-1d9a-4c36-b322-a642073c1057)

later i will write a video-rotation-detection script using this.

**Reply by hexxone** [reply]
Hey, thanks for your continued efforts!

I also think it will take a while to get this feature ready for immich to cover all the edge cases.
Fortunately for me, I had mostly very obvious photos that were wrongly rotated by an older digital camera, and for me, just leaving out the "inconsistent" images is enough for now. I just couldnt be bothered to manually check those >500 images ^^ hence the CLI tool... 

Training a better model to cover the other edge-cases seems like a good idea, I will gladly provide some training time on my 3090 if that helps this awesome community :) I probably just need some help/instructions with downloading & setting up the datasets correctly ^^ 

Just hit me up when you get to the point 👍

**Reply by AllTaken** [reply]
Sounds a lot like you are letting the "perfect" get in the way of of the good.
Having something in Immich that would suggest rotations for images with a decent accuracy, would already be a **_very_** useful addition.
As long as there in general are more accurate suggestions than false positives, I would argue that it is a **_very_** useful feature.
I think in general people have thousands of images. If there's a suggestion to rotate for maybe tens of images, and even 3 of them are correct, it would still be useful, since it basically takes no time to go through tens of images as a human, whereas thousands is basically unfeasible.
Later increments in accuracy would just improve the solution.

**Comment by sildur** [comment]
DigiKam has recently added AI autorotation functionality. The details are in this commit:
https://invent.kde.org/graphics/digikam/-/commit/46002a51837355644ac2f6c0addb6ea4b3a43750

They are using deep-image-orientation-angle-detection.onnx, which can be found here:
https://huggingface.co/Chuckame/deep-image-orientation-angle-detection/tree/main

More details about the model can be found here:
https://github.com/pidahbus/deep-image-orientation-angle-detection

**Reply by Chuckame** [reply]
Amazing! I've started in the past, trying to include the model and corresponding apis in immich, but the python part feels really complex to modify...

**Reply by headtr1ck** [reply]
Looks like a cool model. Only issue I see is that you don't know the certainty of the output. You only get the estimated orientation. But for an image where the orientation is ambiguous you don't know if the output is valid.
I assume that would cause many false positives in large libraries.

A classification version into 4 groups 0, 90, 180, 270 would be better I assume.

**Comment by abanck** [comment]
Seems part of implementing this (and manual rotation tool) is supporting reading image orientation from an XMP sidecar file. I looked at several related issues and PR's and in some this is referenced but could not find an issue/PR directly for reading the XMP file orientation. Is there a related open issue/PR for this, or any plans for this?

Like several other tools [exiftool supports creating XMP sidecar files](https://exiftool.org/metafiles.html) with the [Orientation specification](https://developer.adobe.com/xmp/docs/XMPNamespaces/tiff/) so reading this in Immich would help integration of image rotation detection greatly.

**Comment by headtr1ck** [comment]
Now that rotating is supported in immich I would love to see that feature implemented!
